#454: merge /document into /implement as one configurable orchestrator (#490)
khmelevartem/tether: .claude/skills/implement/SKILL.md

- **BUGFIX root cause** — `bug-reproducer` returns CANNOT REPRODUCE or no hypothesis matches. Mechanics in `bugfix-root-cause`.
- **Cause-vs-issue divergence** — confirmed cause materially diverges from issue body (different mechanism / scope / symptom / severity). Ask: close #N as misdiagnosis and open a new issue, or post a clarifying comment to the issue. Do not silently edit the issue body.
- **Plan conflicts with engineering guides** — surfaced by the `Plan` agent or by recon's living-doc digest, with no clean resolution.
- **Forced-cascade scope expansion** — a change is technically forced but violates an explicit entry in the issue's **Out of scope** section. Ask: fold, split, or re-frame.
- **Smoke red/yellow** — `runtime-verify` verdict is not green after the inner loop.
- **Sub-agent open question** — a sub-agent returns a question it cannot converge on. Relay verbatim, collect answers, re-dispatch the same agent. Routing a contradiction to the owning sub-agent (`spec-writer` / `architect` / `ux-expert`) is automatic — stop at the user only when the sub-agent itself surfaces an unresolvable question.

> - **Final summary** — after all steps converge, commit + push + open PR, then present the PR URL with the summary (AC verdict, smoke verdict if applicable, any `[UNVERIFIABLE]` findings) plus manual test plan if applicable. Do not block on explicit OK before push.

Author: ```suggestion
```

Author: Done in 5e4d56b — the "Final summary" entry is removed from SKILL.md §Gate semantics. It was not a stop condition, it was the end-of-work report; mechanics now live solely in steps.md Step 14 `final-summary`.

---
name: implement
description: Drive a GitHub issue to a review-ready PR end to end — implementation, multi-agent review, and runtime verification — pausing only for decisions that need you. Use to carry any issue (code, docs, or mixed) from start to an opened PR.
---

# /implement — Issue-to-PR orchestrator

You are the orchestrator for a single GitHub issue. You do NOT write code, design artifacts, or review yourself — that is sub-agents' work. You classify, route, aggregate, and gate.

**Goal:** issue → reviewed artifact → open PR, with the user consulted at human-required gates only. Merge is always a manual user decision.

## Input

> Issue number `<N>`.

Author: ```suggestion
Issue number `<N>` — optional.

- **`<N>` provided** — may start fresh work or re-enter an existing PR.
- **`<N>` omitted** — only re-entry is possible. The re-entry gate resolves the issue. If no work started during the session and no open PR - STOP immediately.

```

Author: Applied in 4c1b354 with phrasing aligned to the surrounding voice. `<N>` is now optional; classify.sh resolves the issue from the current branch / open PR on re-entry; STOP when neither resolves an issue.

## How the skill runs

> 1. Run `classify.sh <N>` → mechanical profile facts (`issue`, `reentry`, `pr`, `drift`, `type`, `touched`).

Author: classify дублируется - здесь как отдельный шаг и степ 0 в steps. убирай что то одно.

Author: Done in 4c1b354 — SKILL.md §How runs is now a short paragraph that points to steps.md. The mechanical 4-step list there duplicated Step 0 classify and is gone.

---
name: implement

> description: Issue-to-PR orchestrator for code and docs. Resolves a task profile once (track, type, touched layers), selects which steps run via classify.sh + select-pipeline.sh, then walks steps.md in order. Stops for the user at human-required gates only. Use for any issue — code, docs, or mixed.

Author: description не должен пересказывать механику. он должен отвечать на вопрос: для чего этот скрипт нужно использовать, в каком случае вызвать, что он сделает - в результате, а не в процессе.

Author: Done in caff8d9 — `description` rewritten to state purpose / when-to-use / result; internal mechanics (classify.sh, select-pipeline.sh, "walks steps.md") removed.

## Re-entry contract

The skill is idempotent per issue. `classify.sh` detects PR state and emits `reentry=fresh` or `reentry=pr-feedback`; `select-pipeline.sh` selects the re-entry step set accordingly. Full reconciliation mechanics — drift gate, reading every PR comment, "counted is not read", by-agent attribution routing, and the reply-after-push rule — live in Step 1 (`reentry-reconcile`). The re-entry commit rule (commit into existing branch, no force-push, no new PR) lives in Step 12.

## No-deflection principle

Every demand — from the user or a reviewer — requires a substantive response: either a genuine artifact edit or a justified refusal. Intermediate moves are deflection and are forbidden:

- A reviewer says "remove X" → the coder removes X from the artifact. Adding a KDoc / comment / doc paragraph defending X is not compliance.
- A user question ("is X really needed?" / "why X?") is a request for judgment, not a directive. Answer it; wait for a directive before acting. Silently performing the assumed action is forbidden.

> If the coder responded to "remove X" by documenting X — reject it. Re-dispatch with "change X itself, without justifying it through documentation."

Author: ```suggestion
- A reviewer says "remove X" → the coder removes X from the artifact. Adding a KDoc / comment / doc paragraph defending X is not compliance. If the coder responded to "remove X" by documenting X — reject it. Re-dispatch with "change X itself, without justifying it through documentation."
- A user question ("is X really needed?" / "why X?") is a request for judgment, not a directive. Answer it; wait for a directive before acting. Silently performing the assumed action is forbidden.

```

Author: Applied verbatim in caff8d9.

## Gate semantics — when to stop and ask the user

These MUST-stop gates are **not overridden by session-level autonomy or "skip clarifying questions" hints**. Such hints cover execution-stage trivia within an already-agreed scope; they do not apply here. A one-message pause costs far less than unwinding a unilateral product or architectural decision.

You MUST stop and ask the user:

> - **Spec or AC ambiguity** — issue DoD is missing/stub, FEATURE spec is missing, or spec has blocking open questions. Dispatch `spec-writer` first; stop at the user only if `spec-writer` surfaces questions it cannot converge on, or the issue is non-FEATURE without a DoD.

Author: здесь мы немного поменяем. для реализации фичи обязательно спека. если спеки нет, то задача не готова к выполнению. не запускаем писателя, эскалируем к пользователю. не быть спеки может быть только в задаче, целиком посвященной написанию спеки.

Author: Done in caff8d9 — for a FEATURE, a missing/stub spec now stops at the user (task not ready); `spec-writer` is no longer dispatched as a mitigation, only to resolve blocking open questions in an *existing* spec. (Non-FEATURE missing/stub DoD also stops — made actionable in steps.md `early-gates` in 4c5aeea.)

> - **BUGFIX root cause** — dispatch `bug-reproducer`; stop at the user if CANNOT REPRODUCE or no hypothesis matches. The reproducer must always attempt to observe the symptom even when the cause looks structurally evident.

Author: ```suggestion
- **BUGFIX root cause** — dispatch `bug-reproducer`; stop at the user if CANNOT REPRODUCE or no hypothesis matches. The reproducer must always attempt to observe the symptom even when the cause looks structurally evident or the issue names hypotheses or reasons directly. 
```

Author: Applied verbatim in caff8d9.

> - **Cause-vs-issue divergence** — confirmed cause materially diverges from issue body (different mechanism / scope / symptom / severity). Stop and ask: close #N as misdiagnosis and open a new issue, or rewrite #N body to match the confirmed cause. Do not silently edit the issue body.

Author: rewrite #N body -> post a clarifying comment to issue

Author: Done in caff8d9 — the option 'rewrite #N body' is replaced with 'post a clarifying comment to the issue'; 'do not silently edit the issue body' kept.

> - **Publication of confirmed cause** — once `bug-reproducer` returns a confirmed cause, show the paste-ready block and wait for explicit OK before `gh issue comment`. A team-visible side effect requires a user gate.

Author: а вот это можно удалить. выглядит как место, которое не нуждается в участии пользователя. коммент обязателен

Author: Done in caff8d9 — the user-gate is removed; posting the confirmed cause as an issue comment is now a mandatory step with no stop.

- **Plan ambiguity** — plan conflicts with loaded engineering guides and you have no clean resolution.
- **Forced-cascade scope expansion** — a change is technically forced but falls outside the issue's literal **Out of scope**. Ask: fold, split, or re-frame. Not a heads-up.

> - **Smoke red/yellow** — Step 11 verdict is not green after the inner loop.

Author: кажется, что упоминание Step по номерам теперь не имеет смысла. перепроверь все, поправь

Author: Done — swept every `Step N` cross-reference to step names (ids) across SKILL.md / steps.md / reviewers.md (caff8d9) and the affected docs architecture-principles.md, glossary-discipline.md, review-glossary.md (4c5aeea, 790727e). Note: architecture-principles.md was also factually wrong — it pointed smoke at the old Step 7, which is now the inner-loop; fixed to the `runtime-verify` step.

> - **D1 — Issue framing ambiguity** (docs track) — DoD is missing/stub or contradicts comments irresolvably; requested deliverable is unclear enough that layer classification cannot proceed; a sub-agent returned open questions it cannot converge on — relay verbatim, collect answers, re-dispatch the same agent.

Author: перепиши, чтобы не зависело от того, доки или нет. 
+ это разные гейты
1. requested deliverable is unclear enough that layer classification cannot proceed;
2. a sub-agent returned open questions it cannot converge on — relay verbatim, collect answers, re-dispatch the same agent.

 DoD is missing/stub or contradicts comments irresolvably - уже упомянуто в первом пункте

Author: Done in caff8d9 — rewritten track-agnostic and split into two gates: (1) the deliverable is unclear enough that scope/layer classification cannot proceed; (2) a sub-agent returns open questions it cannot converge on → relay verbatim, re-dispatch. The DoD-missing/stub clause is dropped here (already covered by the Spec/AC gate).

> - **D2 — Cross-doc inconsistency** (docs track) — consistency pass finds a contradiction whose resolution requires a product/technical decision. Route to the owning sub-agent (`spec-writer` / `architect` / `ux-expert`), not directly to the user; stop at the user only if the sub-agent itself surfaces an open question.

Author: тоже не обязательно docs track. 
но тут шиза, ты пишешь в списке явных СТОПОВ про случай, где агент должен сам решить.
сформулируй пункт так, чтобы было только место, в котором надо обратиться именно к юзеру.

Author: Done in caff8d9 — track qualifier dropped; routing a contradiction to the owning sub-agent (spec-writer/architect/ux-expert) is now stated as automatic, not a user STOP. The only user stop is when that sub-agent itself surfaces an open question — so this folded into the sub-agent-open-question gate.

> - **Final summary** — after all steps converge, commit + push + open PR, then present the PR URL with a summary (files changed, AC verdict, smoke verdict if applicable, any `[UNVERIFIABLE]` findings). Do not block on explicit OK before push.

Author: ```suggestion
- **Final summary** — after all steps converge, commit + push + open PR, then present the PR URL with a summary (AC verdict, smoke verdict if applicable, any `[UNVERIFIABLE]` findings) plus manual test plan if applicable. Do not block on explicit OK before push.
```

Author: Applied verbatim in caff8d9 (including 'plus manual test plan if applicable').
